servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -1476,6 +1520,183 @@ public void setState(long fileId, SetStateOptions options)
+    List<Long> workers = Lists.newArrayList();
+    try {
+      for (FileBlockInfo fileBlockInfo : getFileBlockInfoList(fileId)) {
+        for (BlockLocation blockLocation : fileBlockInfo.blockInfo.locations) {
+          workers.add(blockLocation.workerId);
+        }
+      }
+    } catch (FileDoesNotExistException e) {
+      // should not happen
+      throw new RuntimeException(e);
+    } catch (InvalidPathException e) {
+      // should not happen
+      throw new RuntimeException(e);
+    }
+
+    if (workers.size() == 0) {
+      throw new FileDoesNotExistException("The file " + fileId + " does not exist on any worker");
+    } else if (workers.size() > 1) {
+      LOG.info("the file is stored at more than one worker: " + workers);
+    }
+    return workers.get(0);
+  }
+
+  /**
+   * Polls the files to send to the given worker for persistence.
+   *
+   * @param workerId the worker id
+   * @return the list of files
+   * @throws FileDoesNotExistException if the file does not exist
+   * @throws InvalidPathException if the path is invalid
+   */
+  private synchronized List<PersistFile> pollToCheckpoint(long workerId)
+      throws FileDoesNotExistException, InvalidPathException {
+    List<PersistFile> files = Lists.newArrayList();
+    if (!mWorkerToAsyncPersistFile.containsKey(workerId)) {
+      return files;
+    }
+
+    List<Long> toRequestFilePersistence = Lists.newArrayList();
+    Set<Long> scheduledFiles = mWorkerToAsyncPersistFile.get(workerId);
+    for (long fileId : Sets.newHashSet(scheduledFiles)) {
+      InodeFile inode = (InodeFile) mInodeTree.getInodeById(fileId);
+      if (inode.isCompleted()) {
+        toRequestFilePersistence.add(fileId);
+        List<Long> blockIds = Lists.newArrayList();
+        for (FileBlockInfo fileBlockInfo : getFileBlockInfoList(fileId)) {
+          blockIds.add(fileBlockInfo.blockInfo.blockId);
+        }
+
+        PersistFile toCheckpoint = new PersistFile(fileId, blockIds);
+        files.add(toCheckpoint);
+        // update the inode file persisence state
+        inode.setPersistenceState(FilePersistenceState.PERSISTING);
+        scheduledFiles.remove(fileId);
+      }
+    }
+
+    requestFilePersistence(toRequestFilePersistence);
+    // write to journal
+    PersistFilesRequestEntry persistFilesRequest =
+        PersistFilesRequestEntry.newBuilder().addAllFileIds(toRequestFilePersistence).build();
+    writeJournalEntry(
+        JournalEntry.newBuilder().setPersistFilesRequest(persistFilesRequest).build());
+    flushJournal();
+    return files;
+  }
+
+  /**
+   * Request a list of files as being persisted.
+   *
+   * @param fileIds the id of the files
+   * @throws FileDoesNotExistException when a file does not exist
+   */
+  private void requestFilePersistence(List<Long> fileIds) throws FileDoesNotExistException {
+    if (!fileIds.isEmpty()) {
+      LOG.info("Request file persistency: {}", fileIds);
+    }
+    for (long fileId : fileIds) {
```

Comment:
Why does this method not need to synchronize on mInodeTree if it is modifying inode state?